docs/design.md

```diff
@@ -0,0 +1,299 @@
+entries. Store these entries in a Postgres database. Provide a REST API endpoint
+accepting `(ProviderID, PieceCID)` and returning a single `PayloadCID`.
+
+#### Notes
+
+Pieces are immutable. If we receive an advertisement saying that a payload block
+CID was found in a piece CID, then this information remains valid forever, even
+after the SP advertise that they are no longer storing that block. This means
+our indexer can ignore `IsRm` advertisements.
+
+The indexer protocol does not provide any guarantees about the list of CIDs
+advertised for the same Payload CID. Different SPs can advertise different lists
+(e.g. the entries can be ordered differently) or can even cheat and submit CIDs
+that are not part of the Piece. Our indexer must scope the information to each
+index provider.
+
+### Anatomy of IPNI Advertisements
+
+Quoting from
+[Ingestion](https://github.com/ipni/specs/blob/90648bca4749ef912b2d18f221514bc26b5bef0a/IPNI.md#ingestion):
+
+> The indexer reads the advertisement chain starting from the head, reading
+> previous advertisements until a previously seen advertisement, or the end of
+> the chain, is reached. The advertisements and their entries are then processed
+> in order from earliest to head.
+
+An Advertisement has several properties (see
+[the full spec](https://github.com/ipni/specs/blob/90648bca4749ef912b2d18f221514bc26b5bef0a/IPNI.md#advertisements)),
+we need the following ones:
+
+- **`PreviousID`** is the CID of the previous advertisement, and is empty for
+  the 'genesis'.
+
+- **`Metadata`** represents additional opaque data. The metadata for Graphsync
+  retrievals includes the PieceCID that we are looking for.
+
+- **`Entries`** provide a paginated list of multihashes. For our purposes, it's
+  enough to take the first entry and ignore the rest.
+
+Advertisements are made available for consumption by indexer nodes as a set of
+files fetched via HTTP.
+
+Quoting from
+[Advertisement Transfer](https://github.com/ipni/specs/blob/90648bca4749ef912b2d18f221514bc26b5bef0a/IPNI.md#advertisement-transfer):
+
+> All IPNI HTTP requests use the IPNI URL path prefix, `/ipni/v1/ad/`. Indexers
+> and advertisement publishers implicitly use and expect this prefix to precede
+> the requested resource.
+>
+> The IPLD objects of advertisements and entries are represented as files named
+> by their CIDs in an HTTP directory. These files are immutable, so can be
+> safely cached or stored on CDNs. To fetch an advertisement or entries file by
+> CID, the request made by the indexer to the publisher is
+> `GET /ipni/v1/ad/{CID}`.
+
+The IPNI instance running at https://cid.contact provides an API returning the
+list of all index providers from which cid.contact have received announcements:
+
+https://cid.contact/providers
+
+The response provides all the metadata we need to download the advertisements:
+
+- `Publisher.Addrs` describes where we can contact SP's index provider
```

Comment:
I think it doesn't have to be an HTTP address; it could be a Graphsync/libp2p address, too.

However, there is a push from IPNI to move to HTTP transport. For our lighweight indexer, we will require SPs to support the HTTP protocol for handling indexer requests.

Unfortunately, this requires SPs to tweak their Boost configuration and provide the public hostname at which Boost can be reached from outside. On the bright side, I think it's most likely that SPs have to configure this option anyways if they want cid.contact to receive their advertisements, in which case our lightweight indexer is not adding any new requirements.

https://boost.filecoin.io/configuration/http-indexer-announcement

Reply:
I reworded this item as follows:

> `Publisher.Addrs` describes where we can contact SP's index provider to retrieve content for CIDs, e.g., advertisements.